scripts/gb_model/compose_network.py

```diff
@@ -392,6 +392,318 @@ def add_load(
+    )
+    n.add(
+        "Link",
+        ev_dsr.index,
+        suffix=" EV DSR reverse",
+        bus0=ev_dsr.index + " EV",
+        bus1=ev_dsr.index + " EV store",
+        p_nom=ev_dsr["p_nom"].abs(),
+        efficiency=1.0,
+        carrier="EV DSR reverse",
+    )
+
+    # Add EV V2G to the PyPSA network
+    n.add(
+        "Link",
+        ev_v2g.index,
+        suffix=" EV V2G",
+        bus0=ev_v2g.index + " EV store",
+        bus1=ev_v2g.index,
+        p_nom=ev_v2g["p_nom"].abs(),
+        efficiency=1.0,
+        carrier="EV V2G",
+    )
+
+
+def transform_ev_profile_with_shape_adjustment(
+    shape_series, peak_target, annual_target
+):
+    """
+    Transform EV profile to match both peak and annual targets by adjusting the shape.
+
+    This function can squeeze (make peakier) or widen (make flatter) the profile
+    to satisfy both constraints simultaneously.
+
+    Parameters
+    ----------
+    shape_series : pd.Series
+        Normalized EV demand shape (sum = 1)
+    peak_target : float
+        Target peak demand (MW)
+    annual_target : float
+        Target annual energy (MWh)
+
+    Returns
+    -------
+    pd.Series
+        Transformed profile satisfying both constraints
+    """
+    if shape_series.sum() == 0 or len(shape_series) == 0:
+        return pd.Series(
+            [annual_target / len(shape_series)] * len(shape_series),
+            index=shape_series.index,
+        )
+
+    # Normalize input to ensure sum = 1
+    normalized_shape = shape_series / shape_series.sum()
+
+    # Current peak of normalized shape
+    current_peak = normalized_shape.max()
+
+    # Required peak in normalized space
+    required_peak_normalized = peak_target / len(normalized_shape)
+
+    # If the required peak is achievable with current shape, use simple scaling
+    if required_peak_normalized <= current_peak:
+        # Scale entire profile to match annual, then check if peak constraint is satisfied
+        simple_scaled = normalized_shape * annual_target
+        if simple_scaled.max() <= peak_target:
+            return simple_scaled
+
+    # Need to adjust the shape - use power transformation
+    # Higher gamma = more peaked, lower gamma = flatter
+
+    def objective_function(gamma):
+        """Objective function to find optimal shape parameter."""
+        if gamma <= 0:
+            return float("inf")
+
+        # Apply power transformation
+        transformed = normalized_shape**gamma
+        transformed = transformed / transformed.sum()  # Renormalize
+
+        # Scale to match annual target
+        scaled = transformed * annual_target
+
+        # Check how well we satisfy both constraints
+        peak_error = abs(scaled.max() - peak_target) / peak_target
+        annual_error = abs(scaled.sum() - annual_target) / annual_target
+
+        return peak_error + annual_error
+
+    # Find optimal gamma using golden section search
+    from scipy.optimize import minimize_scalar
```

Comment:
Load any dependencies in at the top. Loading them in later makes it harder to keep track of dependencies required by different modules